packages/utils/telemetry-utils/src/perfTelemetryAggregator.ts

```diff
@@ -0,0 +1,93 @@
+    }
+}
+
+/**
+ * Maps a given identifier for a code block that is being subject to telemetry aggregation,
+ * to the runtime measurements for said code block.
+ */
+const statsMap: Map<symbol, Measurements> = new Map<symbol, Measurements>();
+
+/**
+ * Executes the specified code block and keeps track of aggregated telemetry measurements
+ * (number of executions, total duration, min duration, max duration). Writes a telemetry
+ * performance event to the specified logger when the specified number of executions is reached.
+ *
+ * @param codeBlockId - Unique identifier for the code block to be measured.
+ * @param codeToMeasure - The code to be executed and measured.
+ * @param countThreshold - The telemetry performance event with aggregated data will be
+ *                         generated after this many executions of the code block.
+ * @param eventBase - Properties to be included in the telemetry performance
+ *                    event when it is sent.
+ * @param logger - The logger to use to write the telemetry performance event.
+ * @returns Whatever the passed-in code block returns.
+ *
+ * @remarks
+ * It is up to the caller to ensure that the correct symbol is passed in order to aggregate the
+ * same code block correctly. Note that `Symbol('some-id')` called twice will produce different
+ * symbols, and passing those to this function with the same code block will result in two
+ * separate trackings. This helps to prevent collisions between different users of this function,
+ * but also makes it easy for callers to decide if they want to track, for example,
+ * executions from all instances of a given class in a single aggregation, or if each instance
+ * should track separately (keep the symbol as a static property of the class, or an instance
+ * property, respectively).
+ */
+export function ExecuteWithAggregatedTelemetry(
+    codeBlockId: symbol,
+    codeToMeasure: () => any,
+    countThreshold: number,
+    eventBase: any,
+    logger: ITelemetryLogger): any {
+    let stats = statsMap.get(codeBlockId);
+    if (stats === undefined) {
+        stats = new Measurements();
+        statsMap.set(codeBlockId, stats);
+    }
+
+    const start = performance.now();
+    const returnValue = codeToMeasure();
+    const duration = performance.now() - start;
+
+    stats.executionCount++;
+    stats.totalDuration += duration;
+    stats.minDuration = stats.minDuration === undefined
+        ? duration
+        : Math.min(stats.minDuration, duration);
+    stats.maxDuration = stats.maxDuration === undefined
+        ? duration
+        : Math.max(stats.maxDuration, duration);
+
+    if (stats.executionCount >= countThreshold) {
+        const telemetryEvent: ITelemetryPerformanceEvent = {
+            ...eventBase,
+            ...{
+                duration: stats.totalDuration,
+                count: stats.executionCount,
+                minDuration: stats.minDuration,
+                maxDuration: stats.maxDuration,
```

Comment:
I'd suggest not to calculate and report things like min, max, averages or any other math around durations - we can do all of that in Kusto just fine. Just report samples.
One example why any post-processing is hard - I want to be able to understand things like averages, P95 for time distributions. The problem with reporting sum of durations - it's almost impossible from that measure to report P95 of single call. I can only do so if I assume distribution is normal, and then I can do certain math with data in this form to calculate average and standard deviation, and from there P95. It's a lot of effort and not exact.

If you were reporting just the actual measurement every 100th call (or so), I can just use functions built into Kusto to do all this math exactly.

Reply:
> I'd suggest not to calculate and report things like min, max, averages or any other math around durations - we can do all of that in Kusto just fine. Just report samples.

I talked with @anthony-murphy about this and I leaned towards aggregation instead of sampling to guarantee that we wouldn't miss outliers which would potentially drive our max/percentiles way up, and in principle could be missed if we just sample. With that mindset, at least min and max seemed worth keeping so in Kusto we can also do min/max over those fileds and get the actual min/max over all data points instead of just over the samples. We were aware that percentiles over the raw data become impossible to calculate in this case, and at most we would be able to look at percentiles of the max duration across aggregated batches of measurements.

If we feel confident that we can get a decent enough picture purely out of telemetry volume, even if we're potentially losing the outliers here or there, I'm happy to do sampling instead (and I understand `duration` is enough in that case).

Reply:
i wonder if we could do both. could the event we log be both an aggregation of that data until that point, but also the specific data for the calls that triggers the log?

Reply:
i would probably make the sample the default, but add  an aggDuration, and aggCount columns to track the aggregate data. this is relatively cheap to do, and a for future users has the nice property that they don't need to choose between sampling and aggregation.